// localforage-driver-session-storage/Store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stores = {};
/** @internal */
var Store = /** @class */ (function () {
    function Store(kp) {
        this.kp = kp;
    }
    Store.resolve = function (kp) {
        if (!stores[kp]) {
            stores[kp] = new Store(kp);
        }
        return stores[kp];
    };
    Store.prototype.clear = function () {
        var keys = this.keys();
        for (var i = 0; i < keys.length; i++) {
            this.rm(keys[i]);
        }
    };
    Store.prototype.drop = function () {
        this.clear();
        delete stores[this.kp];
    };
    Store.prototype.get = function (key) {
        return sessionStorage.getItem(this.kp + key);
    };
    Store.prototype.key = function (idx) {
        return this.keys()[idx];
    };
    Store.prototype.keys = function () {
        var out = [];
        var key;
        for (var i = sessionStorage.length - 1; i >= 0; i--) {
            key = sessionStorage.key(i);
            if (key && key.indexOf(this.kp) === 0) {
                out.push(key.substr(this.kp.length));
            }
        }
        return out;
    };
    Store.prototype.rm = function (key) {
        sessionStorage.removeItem(this.kp + key);
    };
    Store.prototype.set = function (k, v) {
        sessionStorage.setItem(this.kp + k, v);
    };
    return Store;
}());
exports.Store = Store;
